compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m

```diff
@@ -52,176 +25,101 @@ @interface UIViewController(CMPUIKitUtilsPrivate)
-        [self.parentViewController.cmp_uikit_utils_deinitMonitor filterUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(CMPDeinitNotifier *notifier, NSDictionary<NSString *,id> * _Nullable bindings) {
-            return notifier != self.cmp_parentDeinitNotifier;
-        }]];
-    }
-
     [super didMoveToParentViewController:parent];
-
-    if (parent != nil && !self.cmp_viewControllerInWindowHierarchy) {
-        [self cmp_verifyHierarchyAndScheduleNotificationIfNeeded];
+    
+    if (parent) {
+        [self transitLifecycleToStarted];
     }
-
-    [self cmp_setNeedsVerifyHierarchy];
-}
-
-- (void)removeFromParentViewController {
-    [super removeFromParentViewController];
-
-    [self cmp_setNeedsVerifyHierarchy];
 }
 
-- (void)viewDidLoad {
-    if ([self.view isKindOfClass:[CMPView class]]) {
-        __weak typeof(self) weakSelf = self;
-        ((CMPView *)self.view).windowDidChangeBlock = ^{
-            [weakSelf cmp_setNeedsVerifyHierarchy];
-        };
-    } else {
-        NSAssert(false, @"The view of CMPViewController should be either CMPView or subclassed from CMPView");
+/// This method can be triggered by both `viewWillAppear` and `didMoveToParentViewController`. Mind that in second case the `view` could still not be constructed (adding `UIViewController`  to inactive tab of `UITabBarController`, for example), so the `ComposeScene` is not neccessarily created at this point.
+- (void)transitLifecycleToStarted {
+    switch (_lifecycleState) {
+        case CMPViewControllerLifecycleStateDestroyed:
+            @throw [NSException exceptionWithName:@"CMPViewControllerMisuse"
+                                           reason:@"CMPViewController shouldn't be reused after completely removed from hierarchy, because it's logically marked as Destroyed. You must create a new CMPViewController and use it instead."
+                                         userInfo:nil];
+        case CMPViewControllerLifecycleStateInitalized:
+            _lifecycleState = CMPViewControllerLifecycleStateStarted;
+            [self viewControllerDidEnterWindowHierarchy];
+            [self scheduleHierarchyContainmentCheck];
+            break;
+        case CMPViewControllerLifecycleStateStarted:
+            break;
     }
-
-    [super viewDidLoad];
-}
-
-- (void)loadView {
-    self.view = [CMPView new];
 }
 
-- (void)cmp_setNeedsVerifyHierarchy {
-    if (self.cmp_needsVerifyHierarchy) {
-        return;
-    }
-    self.cmp_needsVerifyHierarchy = true;
-
-    __weak typeof(self) weakSelf = self;
-    dispatch_async(dispatch_get_main_queue(), ^{
-        weakSelf.cmp_needsVerifyHierarchy = false;
-        [weakSelf cmp_verifyHierarchyAndScheduleNotificationIfNeeded];
+- (void)scheduleHierarchyContainmentCheck {
+    double delayInSeconds = 0.5;
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{        
```

Comment:
Please don't forget to use weak `self`

Reply:
I don't think it's needed here. `self` is captured by a temporary block to be executed on the main dispatch queue later, since we don't own this queue in any way, there is no retain cycle.

Reply:
Ah, I see. Ok!